File: src/server/tools/searchAuthors.ts
import { z } from "zod";

import type { Catalog } from "../catalog.js";
import type { Helper } from "../helper.js";

export const searchAuthorsInput = {
    query: z.string().describe("Arabic search phrase, matched against author name + biography."),
    max_results: z
        .number()
        .int()
        .min(1)
        .max(100)
        .default(20)
        .describe("Maximum number of hits to return (1-100, default 20)."),
};

interface RawHit {
    author_id: number;
    snippet: string;
}

interface RawEnvelope {
    total_hits: number;
    returned: number;
    query: string;
    normalized_tokens: string[];
    results: RawHit[];
}

export interface SearchAuthorHit {
    author_id: number;
    author_name: string;
    death_year: number | null;
    snippet: string;
}

export interface SearchAuthorsOutput {
    total_hits: number;
    returned: number;
    query: string;
    normalized_tokens: string[];
    results: SearchAuthorHit[];
}

export async function searchAuthors(
    helper: Helper,
    catalog: Catalog,
    args: { query: string; max_results: number },
): Promise<SearchAuthorsOutput> {
    const raw = await helper.request<RawEnvelope>("search_authors", args);
    const enriched: SearchAuthorHit[] = raw.results.map((hit) => {
        const author = catalog.author(hit.author_id);
        return {
            author_id: hit.author_id,
            author_name: author.author_name,
            death_year: author.death_year,
            snippet: hit.snippet,
        };
    });
    return {
        total_hits: raw.total_hits,
        returned: raw.returned,
        query: raw.query,
        normalized_tokens: raw.normalized_tokens,
        results: enriched,
    };
}

File: src/java/src/main/java/ws/shamela/mcp/SearchAuthors.java
package ws.shamela.mcp;

import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.apache.lucene.document.Document;
import org.apache.lucene.index.StoredFields;
import org.apache.lucene.index.Term;
import org.apache.lucene.search.BooleanClause;
import org.apache.lucene.search.BooleanQuery;
import org.apache.lucene.search.IndexSearcher;
import org.apache.lucene.search.Query;
import org.apache.lucene.search.ScoreDoc;
import org.apache.lucene.search.TermQuery;
import org.apache.lucene.search.TopDocs;

/**
 * shamela_search_authors — match Arabic tokens against the pre-built `author`
 * index's `body` field (author name + biography). Snippets from `body_store`.
 * Returns minimal hits; Node fills in display names + Hijri death years.
 */
public final class SearchAuthors {

    private SearchAuthors() {}

    static final String INDEX = "author";

    public static Map<String, Object> run(
            IndexCache indexCache,
            String rawQuery,
            int maxResults
    ) throws IOException {
        List<String> tokens = Normalize.normalizeQuery(rawQuery);
        Map<String, Object> envelope = new LinkedHashMap<>();
        envelope.put("query", rawQuery == null ? "" : rawQuery);
        envelope.put("normalized_tokens", tokens);

        if (tokens.isEmpty()) {
            envelope.put("total_hits", 0);
            envelope.put("returned", 0);
            envelope.put("results", List.of());
            return envelope;
        }

        BooleanQuery.Builder b = new BooleanQuery.Builder();
        for (String tok : tokens) {
            b.add(new TermQuery(new Term("body", tok)), BooleanClause.Occur.MUST);
        }
        Query q = b.build();

        IndexSearcher searcher = indexCache.searcher(INDEX);
        StoredFields stored = indexCache.storedFields(INDEX);

        int safeMax = Math.max(1, Math.min(maxResults, 100));
        long total = searcher.count(q);
        TopDocs top = searcher.search(q, safeMax);

        List<Map<String, Object>> results = new ArrayList<>(top.scoreDocs.length);
        for (ScoreDoc sd : top.scoreDocs) {
            Document doc = stored.document(sd.doc);
            String idField = doc.get("id");
            if (idField == null) continue;
            int authorId;
            try {
                authorId = Integer.parseInt(idField.trim());
            } catch (NumberFormatException e) {
                continue;
            }

            String biography = nullToEmpty(doc.get("body_store"));
            String snippet = !biography.isEmpty() ? Snippet.make(biography, tokens) : "";

            Map<String, Object> hit = new LinkedHashMap<>();
            hit.put("author_id", authorId);
            hit.put("snippet", snippet);
            results.add(hit);
        }

        envelope.put("total_hits", (int) Math.min(total, Integer.MAX_VALUE));
        envelope.put("returned", results.size());
        envelope.put("results", results);
        return envelope;
    }

    private static String nullToEmpty(String s) {
        return s == null ? "" : s;
    }
}
